lib/rubocop/cop/rspec/rails/inferred_spec_type.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      module Rails
+        # Identifies redundant spec type.
+        #
+        # @safety
+        #   This cop is marked as unsafe because
+        #   `config.infer_spec_type_from_file_location!` may not be enabled.
+        #
+        # @example
+        #
+        #   # bad
+        #   # spec/models/user_spec.rb
+        #   RSpec.describe User, type: :model
+        #
+        #   # good
+        #   # spec/models/user_spec.rb
+        #   RSpec.describe User
+        #
+        #   # good
+        #   # spec/models/user_spec.rb
+        #   RSpec.describe User, type: :request
+        #
+        class InferredSpecType < Base
+          extend AutoCorrector
+
+          # @return [Array<Hash>]
+          INFERENCES = [
+            { prefix: 'spec/channels/', type: :channel },
+            { prefix: 'spec/controllers/', type: :controller },
+            { prefix: 'spec/features/', type: :feature },
+            { prefix: 'spec/generator/', type: :generator },
+            { prefix: 'spec/helpers/', type: :helper },
+            { prefix: 'spec/jobs/', type: :job },
+            { prefix: 'spec/mailboxes/', type: :mailbox },
+            { prefix: 'spec/mailers/', type: :mailer },
+            { prefix: 'spec/models/', type: :model },
+            { prefix: 'spec/requests/', type: :request },
+            { prefix: 'spec/integration/', type: :request },
+            { prefix: 'spec/api/', type: :request },
+            { prefix: 'spec/routing/', type: :routing },
+            { prefix: 'spec/system/', type: :system },
+            { prefix: 'spec/views/', type: :view }
+          ].freeze
+
+          MSG = 'Remove redundant spec type.'
+
+          RESTRICT_ON_SEND = %i[
+            describe
+          ].freeze
+
+          # @param [RuboCop::AST::SendNode] node
+          def on_send(node)
+            pair_node = describe_with_type(node)
+            return unless pair_node
+            return unless inferred_type?(pair_node)
+
+            removable_node = detect_removable_node(pair_node)
+            add_offense(removable_node) do |corrector|
+              autocorrect(corrector, removable_node)
+            end
+          end
+
+          private
+
+          # @!method describe_with_type(node)
+          #   @param [RuboCop::AST::SendNode] node
+          #   @return [RuboCop::AST::PairNode, nil]
+          def_node_matcher :describe_with_type, <<~PATTERN
+            (send
+              { (const nil? :RSpec) | nil? }
+              _
+              _
+              _*
+              ({ hash | kwargs }
+                (pair ...)*
+                $(pair (sym :type) (sym _))
+                (pair ...)*
+              )
```

Comment:
I think that this can be simplified down to `(hash <$(pair (sym :type) sym) ...>))` with [match in any order node pattern feature](https://docs.rubocop.org/rubocop-ast/node_pattern.html#for-match-in-any-order) as we do in e.g. lib/rubocop/cop/rspec/focus.rb

Reply:
I've simplified the pattern slightly